File: C++/532.cpp
__________________________________________________________________________________________________
sample 16 ms submission
auto speedup =[]() {
    std::ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return nullptr;
}();

class Solution {
public:
    int findPairs(vector<int>& nums, int k) 
    {
        if (k < 0)
            return 0;
        int count = 0;
        if (k == 0) {
            unordered_map<int,int> xs;
            for (auto n : nums)
                xs[n]++;
            for (auto& x : xs)
                if (x.second > 1)
                    ++count;
        } else {
            unordered_set<int> xs(nums.begin(), nums.end());
            for (auto x : xs) 
                if (xs.find(x+k) != xs.end())
                    ++count;
        }
        return count;
    }
};
__________________________________________________________________________________________________
sample 9816 kb submission
static const auto speedup = []() { 
    std::ios::sync_with_stdio(false); 
    std::cin.tie(nullptr); 
    return 0; 
}();
class Solution {
public:
    
    
    int findPairs(vector<int>& nums, int k) {
        int cnt = 0;
        sort(nums.begin(), nums.end());
        
        for(int left = 0; left < min(nums.size(), nums.size() - 1); left++) {
            
            if(left && nums[left-1] == nums[left]) continue; // avoid duplicates
            
            int right = left + 1;
            
            //cout << "left: " << left << " right: " << right << endl;
            
            while(right < nums.size() && nums[right] - nums[left] < k) {
                right++;
            }
            if(right < nums.size() && nums[right] - nums[left] == k) cnt++;
        }        
        return cnt;
    }
};
__________________________________________________________________________________________________

File: Java/532.java
__________________________________________________________________________________________________
sample 3 ms submission
class Solution {
    public int findPairs(int[] nums, int k) {
        if (k < 0) {
            return 0;
        }

        Arrays.sort(nums);

        int left = 0, right = 1, count = 0;

        while (right < nums.length) {
            if (nums[right] - nums[left] > k) {
                left++;
            } else if (left == right || nums[right] - nums[left] < k) {
                right++;
            } else {
                count++;
                left++;
                right++;
                while (right < nums.length && nums[right] == nums[right - 1]) {
                    right++;
                }
            }
        }

        return count;
    }
}
__________________________________________________________________________________________________
sample 38520 kb submission
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

class Solution {
    public int findPairs(int[] nums, int k) {
        if (k < 0) {
             return 0;
        }
        Map<Integer, Integer> map = new HashMap<>();
        for (int n : nums) {
            map.compute(n + k, (key, value) -> value == null ? 1 : value + 1);
        }
        Set<List<Integer>> set = new HashSet<>();
        for (int num : nums) {
            Integer value = map.get(num);
            if (value != null) {
                int a = Math.min(num, num - k);
                int b = Math.max(num, num - k);
                if (k == 0) {
                    if (value > 1) {
                        set.add(Arrays.asList(num, num));
                    }
                } else {
                    set.add(Arrays.asList(a, b));
                }
            }
        }
        return set.size();
    }
}
__________________________________________________________________________________________________
